File: server/src/services/dataAnonymizer.js
/**
 * Data Anonymization Service
 * Removes/masks sensitive information before sending to OpenAI
 */

// Common patterns to detect and anonymize
const PATTERNS = {
  // Email addresses
  email: /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b/gi,
  
  // Phone numbers (various formats)
  phone: /(\+?\d{1,3}[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}\b/g,
  
  // URLs
  url: /https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_\+.~#?&//=]*)/gi,
  
  // Social Security Numbers (US)
  ssn: /\b\d{3}-\d{2}-\d{4}\b/g,
  
  // Addresses (basic pattern)
  address: /\b\d{1,5}\s+([A-Z][a-z]+\s+){1,3}(Street|St|Avenue|Ave|Road|Rd|Boulevard|Blvd|Lane|Ln|Drive|Dr|Court|Ct|Circle|Cir|Way)\b/gi,
  
  // Salary ranges
  salary: /\$\s*\d{1,3}(,\d{3})*(\s*-\s*\$?\s*\d{1,3}(,\d{3})*)?(\s*(per|\/)\s*(year|yr|annum|hour|hr|month|mo))?/gi,
  
  // Specific company names (will be populated dynamically)
  companyName: null,
};

// Sensitive keywords that might reveal company strategy
const SENSITIVE_KEYWORDS = [
  'confidential',
  'proprietary',
  'internal only',
  'do not share',
  'nda required',
  'trade secret',
  'patent pending',
  'unreleased',
  'stealth mode',
];

/**
 * Anonymize text by removing/masking sensitive information
 * @param {string} text - Original text
 * @param {Object} options - Anonymization options
 * @returns {Object} { anonymizedText, removedItems, warnings }
 */
export function anonymizeText(text, options = {}) {
  if (!text) return { anonymizedText: '', removedItems: [], warnings: [] };

  const {
    maskEmails = true,
    maskPhones = true,
    maskUrls = true,
    maskSalaries = true,
    maskAddresses = true,
    maskSSN = true,
    companyName = null,
    removeConfidential = true,
  } = options;

  let anonymized = text;
  const removedItems = [];
  const warnings = [];

  // 1. Mask emails
  if (maskEmails) {
    const emails = text.match(PATTERNS.email) || [];
    emails.forEach((email, i) => {
      const domain = email.split('@')[1];
      const masked = `[EMAIL_${i + 1}@${domain}]`;
      anonymized = anonymized.replace(email, masked);
      removedItems.push({ type: 'email', original: email, masked });
    });
  }

  // 2. Mask phone numbers
  if (maskPhones) {
    const phones = text.match(PATTERNS.phone) || [];
    phones.forEach((phone, i) => {
      const masked = `[PHONE_${i + 1}]`;
      anonymized = anonymized.replace(phone, masked);
      removedItems.push({ type: 'phone', original: phone, masked });
    });
  }

  // 3. Mask URLs (but keep domain for context)
  if (maskUrls) {
    const urls = text.match(PATTERNS.url) || [];
    urls.forEach((url, i) => {
      try {
        const domain = new URL(url).hostname;
        const masked = `[URL_${i + 1}: ${domain}]`;
        anonymized = anonymized.replace(url, masked);
        removedItems.push({ type: 'url', original: url, masked });
      } catch (e) {
        // Invalid URL, skip
      }
    });
  }

  // 4. Mask SSN
  if (maskSSN) {
    const ssns = text.match(PATTERNS.ssn) || [];
    ssns.forEach((ssn, i) => {
      const masked = `[SSN_${i + 1}]`;
      anonymized = anonymized.replace(ssn, masked);
      removedItems.push({ type: 'ssn', original: ssn, masked });
    });
  }

  // 5. Mask addresses
  if (maskAddresses) {
    const addresses = text.match(PATTERNS.address) || [];
    addresses.forEach((addr, i) => {
      const masked = `[ADDRESS_${i + 1}]`;
      anonymized = anonymized.replace(addr, masked);
      removedItems.push({ type: 'address', original: addr, masked });
    });
  }

  // 6. Mask salary information
  if (maskSalaries) {
    const salaries = text.match(PATTERNS.salary) || [];
    salaries.forEach((salary, i) => {
      const masked = `[SALARY_RANGE_${i + 1}]`;
      anonymized = anonymized.replace(salary, masked);
      removedItems.push({ type: 'salary', original: salary, masked });
    });
  }

  // 7. Mask specific company name
  if (companyName) {
    const companyRegex = new RegExp(`\\b${escapeRegex(companyName)}\\b`, 'gi');
    const matches = text.match(companyRegex) || [];
    if (matches.length > 0) {
      anonymized = anonymized.replace(companyRegex, '[COMPANY_NAME]');
      removedItems.push({ type: 'company', original: companyName, masked: '[COMPANY_NAME]' });
    }
  }

  // 8. Check for confidential keywords
  if (removeConfidential) {
    const lowerText = text.toLowerCase();
    SENSITIVE_KEYWORDS.forEach((keyword) => {
      if (lowerText.includes(keyword)) {
        warnings.push(`Contains potentially confidential keyword: "${keyword}"`);
      }
    });
  }

  // 9. Remove lines containing "confidential" or "internal"
  if (removeConfidential) {
    const lines = anonymized.split('\n');
    const filteredLines = lines.filter((line) => {
      const lower = line.toLowerCase();
      return !SENSITIVE_KEYWORDS.some((kw) => lower.includes(kw));
    });
    
    if (filteredLines.length < lines.length) {
      warnings.push(`Removed ${lines.length - filteredLines.length} lines containing sensitive keywords`);
      anonymized = filteredLines.join('\n');
    }
  }

  return {
    anonymizedText: anonymized.trim(),
    removedItems,
    warnings,
    isAnonymized: removedItems.length > 0 || warnings.length > 0,
  };
}

/**
 * Anonymize resume data before sending to OpenAI
 */
export function anonymizeResume(resumeText, options = {}) {
  return anonymizeText(resumeText, {
    maskEmails: true,
    maskPhones: true,
    maskUrls: true,
    maskSalaries: false, // Keep salary expectations in resumes
    maskAddresses: true,
    maskSSN: true,
    removeConfidential: false, // Resumes rarely have confidential info
    ...options,
  });
}

/**
 * Anonymize job description before sending to OpenAI
 */
export function anonymizeJobDescription(jobText, companyName, options = {}) {
  return anonymizeText(jobText, {
    maskEmails: true,
    maskPhones: true,
    maskUrls: true,
    maskSalaries: true, // Mask internal salary ranges
    maskAddresses: true,
    maskSSN: false,
    companyName,
    removeConfidential: true, // Remove confidential sections
    ...options,
  });
}

/**
 * Create a mapping to restore anonymized data (if needed)
 */
export function createRestoreMap(removedItems) {
  const map = {};
  removedItems.forEach((item) => {
    map[item.masked] = item.original;
  });
  return map;
}

/**
 * Restore anonymized text using the mapping
 */
export function restoreText(anonymizedText, restoreMap) {
  let restored = anonymizedText;
  Object.entries(restoreMap).forEach(([masked, original]) => {
    restored = restored.replace(new RegExp(escapeRegex(masked), 'g'), original);
  });
  return restored;
}

/**
 * Escape special regex characters
 */
function escapeRegex(str) {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

/**
 * Get anonymization statistics
 */
export function getAnonymizationStats(result) {
  const stats = {
    totalRemoved: result.removedItems.length,
    byType: {},
    warnings: result.warnings.length,
  };

  result.removedItems.forEach((item) => {
    stats.byType[item.type] = (stats.byType[item.type] || 0) + 1;
  });

  return stats;
}

export default {
  anonymizeText,
  anonymizeResume,
  anonymizeJobDescription,
  createRestoreMap,
  restoreText,
  getAnonymizationStats,
};

File: server_py/services/data_anonymizer.py
"""
services/data_anonymizer.py — PII scrubbing before sending data to OpenAI.

Detects and masks sensitive information using regex patterns:
- Emails, phone numbers, SSNs, street addresses, salary figures, URLs
- Optional company name redaction
- Strips lines containing confidential/proprietary keywords

Two convenience wrappers:
- anonymize_resume()           — keeps salary expectations, skips confidential filter
- anonymize_job_description()  — masks salary ranges, removes confidential lines
"""

import re

PATTERNS = {
    "email":   re.compile(r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b"),
    "phone":   re.compile(r"(\+?\d{1,3}[.\-\s]?)?\(?\d{3}\)?[.\-\s]?\d{3}[.\-\s]?\d{4}\b"),
    "ssn":     re.compile(r"\b\d{3}-\d{2}-\d{4}\b"),
    "address": re.compile(r"\b\d{1,5}\s+(?:[A-Z][a-z]+\s+){1,3}(?:Street|St|Avenue|Ave|Road|Rd|Boulevard|Blvd|Lane|Ln|Drive|Dr|Court|Ct)\b", re.I),
    "salary":  re.compile(r"\$\s*\d{1,3}(?:,\d{3})*(?:\s*-\s*\$?\s*\d{1,3}(?:,\d{3})*)?(?:\s*(?:per|/)\s*(?:year|yr|month|hour|hr))?", re.I),
    "url":     re.compile(r"https?://(?:www\.)?[-a-zA-Z0-9@:%._+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b[-a-zA-Z0-9()@:%_+.~#?&/=]*", re.I),
}

SENSITIVE_KEYWORDS = ["confidential","proprietary","internal only","do not share","nda required","trade secret"]


def anonymize_text(text: str, **opts) -> dict:
    if not text:
        return {"anonymizedText": "", "removedItems": [], "warnings": []}

    mask_emails  = opts.get("maskEmails", True)
    mask_phones  = opts.get("maskPhones", True)
    mask_urls    = opts.get("maskUrls", True)
    mask_ssn     = opts.get("maskSSN", True)
    mask_addr    = opts.get("maskAddresses", True)
    mask_salary  = opts.get("maskSalaries", True)
    company      = opts.get("companyName")
    remove_conf  = opts.get("removeConfidential", True)

    result = text
    removed = []
    warnings = []

    def _mask(pat_key, label, enabled):
        nonlocal result
        if not enabled:
            return
        for i, m in enumerate(PATTERNS[pat_key].finditer(text), 1):
            masked = f"[{label}_{i}]"
            result = result.replace(m.group(0), masked)
            removed.append({"type": pat_key, "original": m.group(0), "masked": masked})

    _mask("email",   "EMAIL",   mask_emails)
    _mask("phone",   "PHONE",   mask_phones)
    _mask("ssn",     "SSN",     mask_ssn)
    _mask("address", "ADDRESS", mask_addr)
    _mask("salary",  "SALARY",  mask_salary)

    if mask_urls:
        for i, m in enumerate(PATTERNS["url"].finditer(text), 1):
            try:
                from urllib.parse import urlparse
                domain = urlparse(m.group(0)).netloc
                masked = f"[URL_{i}: {domain}]"
            except Exception:
                masked = f"[URL_{i}]"
            result = result.replace(m.group(0), masked)
            removed.append({"type": "url", "original": m.group(0), "masked": masked})

    if company:
        pat = re.compile(rf"\b{re.escape(company)}\b", re.I)
        if pat.search(text):
            result = pat.sub("[COMPANY_NAME]", result)
            removed.append({"type": "company", "original": company, "masked": "[COMPANY_NAME]"})

    lower = text.lower()
    for kw in SENSITIVE_KEYWORDS:
        if kw in lower:
            warnings.append(f'Contains potentially confidential keyword: "{kw}"')

    if remove_conf:
        lines = result.split("\n")
        filtered = [l for l in lines if not any(kw in l.lower() for kw in SENSITIVE_KEYWORDS)]
        if len(filtered) < len(lines):
            warnings.append(f"Removed {len(lines) - len(filtered)} lines with sensitive keywords")
            result = "\n".join(filtered)

    return {
        "anonymizedText": result.strip(),
        "removedItems": removed,
        "warnings": warnings,
        "isAnonymized": bool(removed or warnings),
    }


def anonymize_resume(text: str, **opts) -> dict:
    return anonymize_text(text, maskEmails=True, maskPhones=True, maskUrls=True,
                          maskSalaries=False, maskAddresses=True, maskSSN=True,
                          removeConfidential=False, **opts)


def anonymize_job_description(text: str, company_name: str = None, **opts) -> dict:
    return anonymize_text(text, maskEmails=True, maskPhones=True, maskUrls=True,
                          maskSalaries=True, maskAddresses=True, maskSSN=False,
                          companyName=company_name, removeConfidential=True, **opts)
